### PR TITLE
feat(gatsby): Schema rebuilding

### LINKED ISSUE #18939: GraphQL Schema rebuilding
Currently in develop the schema doesn't change when data changes, even thought this can sometimes happen. This proposes adding updating schema when needed.

Lots of the ground work has already been done, things that should react to schema changes do react to them, thus only implementing actual schema updating would be required.

We already rebuild a schema during our bootstrap process due to context inference (however we will disable querying context this way in v3). As a side effect of this PR we can put deprecation warning for this. See [RFC about removing context queryablity](https://github.com/gatsbyjs/rfcs/pull/26)

The basic process should collect node ids and types that change by subscribing to node redux events, then when API_RUNNING_QUEUE_EMPTY is triggered, it should call some kind of schema rebuilding function with those ids/types. Schema rebuilding should compare new nodes example object with an example object that is cached for the previous run (can be maybe metadata stored in type composer) and perform updates to schema in case this happened. It should then recreate the schema.

### PR DESCRIPTION
**Pull request by vladar**
## Description

Enables schema rebuilding so that `gatsby develop` restart is not required when nodes change. It should also help with incremental builds.

This whole feature is related to types and fields created via inference. Types and fields created via schema customization are considered static and shouldn't rebuild.

## TODO

- [x] Initial PoC
- [x] Remove stale inferred fields
- [x] Handle `ADD_FIELD_TO_NODE` in the reducer
- [x] Rework existing tests to use new APIs (utilizing metadata vs. list of nodes for inferring)
- [x] Derived types (filter/sort types, nested inferred types, etc.)
- [x] Root fields / arguments
- [x] Handle implicit parent-child relations and `ADD_CHILD_NODE_TO_PARENT_NODE`
- [x] Updating extensions for node and derived types
- [x] New tests for this specific feature (relied on existing tests so far which are green)
- [x] Make sure types and fields created via schema customization are not rebuilt
- [x] Move schema customization before node sourcing (to skip inference for types marked with `@dontInfer`)
- [x] Parse type defs before dispatching CREATE_TYPES action
- [x] Fix eslint to pick up new schema after rebuild
- [x] Re-run queries after rebuild (as they may be broken after schema change)
- [x] Improve dirty-checks to minimize false-positives
- [ ] Test with multiple medium to big size projects

## Use cases supported in `develop`

1. Adding new node types
2. Adding nodes with new or structurally modified fields
3. Modifying nodes (adding new fields, changing field types, etc.)
4. Deleting fields from node types
5. Deleting of a type when there are no nodes or fields that could produce it in the first place

In these cases, schema rebuilds without restarting. Non-structural node changes do not trigger a rebuild (i.e., nodes added with the same structure)

## Inference Refactoring
These changes required refactoring of the inference process (or more specifically - `getExampleValue`).

**Before** this change, `getExampleValue` was looping through all the nodes of a given type to construct an example value, which was later utilized by type inference. This approach doesn't play well with incremental schema rebuilding because it requires `O(N*M)` to create example value (where `N` is the number of node fields, including nested fields, and `M` is the number of nodes of the given type)

**After** this PR `getExampleValue` uses node metadata stored in the redux store and updated on every node-related action. Initial metadata building during bootstrap is the same `O(N*M)` but `getExampleValue` is now `O(N)` which makes it reasonably fast for incremental re-building of the example value.

### Caveat: conflict tracking
Conflict tracking for arrays is tricky, i.e.: `{ a: [5, "foo"] }` and `{ a: [5] }, { a: ["foo"] }`
are represented identically in metadata and reported identically. To workaround it we additionally track first NodeId for a type:
```js
{ a: { array: { item: { int: { total: 1, first: "1" }, string: { total: 1, first: "1" } }}
{ a: { array: { item: { int: { total: 1, first: "1" }, string: { total: 1, first: "2" } }}
```
This helps producing more useful conflict reports (still rare edge cases possible when reporting may be confusing, i.e. when node is deleted)

### Caveat: dirty checking
Some plugins will delete nodes and then re-create on any change. So even if final metadata is identical it will be still marked as `dirty`. We additionally compare dirty metadata between calls (and skip rebuilding if the inferred structure is the same)

### Caveat: rebuild granularity
Currently, we rebuild full schema instance from scratch (only when there are some structural changes). Granular updates are complicated as `graphql-compose` wasn't designed for mutations like this: it is great for complex schema builds but provides little help when you delete or modify types / fields / arguments.

## Possible follow-ups

* Rebuild granularity (see caveat above). This will likely require some coordination with `graphql-compose` author or our own type/field/arg dependency tracking. But the complexity involved might not worth the effort.

* We could use metadata directly for inference (vs. `example value`). Metadata has more information for inference or granular rebuilds. For example, we could handle data conflicts more gently. Say when  < 1% of nodes have a conflicting field type - add the majority field and warn about specific conflicting node id (currently we remove a field on conflicts and report both nodes)

* Investigate webpack invalidation on schema change, see https://github.com/gatsbyjs/gatsby/pull/19092#pullrequestreview-316051533

## Related Issues

Fixes #18939

### CODE REVIEW

File: packages/gatsby/src/utils/node-descriptor.js [hunk 1]
```diff
@@ -0,0 +1,346 @@
+/*
```
Comment (freiksenet) [inline packages/gatsby/src/utils/node-descriptor.js]: Nice!

File: packages/gatsby/src/utils/node-descriptor.js [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+  console.log(meta.fieldMap)
+  // outputs: {
+  //   foo: {
+  //     int: { total: 1, example: 25 },
+  //     string: { total: 1, example: 'conflict' },
+  //   },
+  //   bar: {
+  //     string: { total: 1, example: 'str' },
+  //   },
+  // }
```
Comment (pieh) [inline packages/gatsby/src/utils/node-descriptor.js]: How are nested objects represented here? How arrays are represented?

I.e. 
what would
```
const node2 = { id: '1', nested: { foo: 'bar' }, array: [1, 2, 3, "string"] }
```
produce

(if this is handled already)

Comment (pieh) [reply packages/gatsby/src/utils/node-descriptor.js]: Ah there are types below - ignore above ;)

Comment (vladar) [reply packages/gatsby/src/utils/node-descriptor.js]: Yeah, those are just simple usage examples, didn't want to distract with all the recursion stuff here %)

File: packages/gatsby/src/bootstrap/schema-hot-reloader.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
+const { debounce } = require(`lodash`)
+const { emitter, store } = require(`../redux`)
+const { rebuildWithTypes } = require(`../schema`)
+const report = require(`gatsby-cli/lib/reporter`)
+
+// API_RUNNING_QUEUE_EMPTY could be emitted multiple types
+// in a short period of time, so debounce seems reasonable
```
[the remark this answers is not in the document]
Comment (vladar) [reply packages/gatsby/src/bootstrap/schema-hot-reloader.js]: Oops, missed this comment, sorry. It makes sense. But I suggest doing this in a separate PR

File: packages/gatsby/src/schema/infer/inference-metadata.js [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+
+```javascript
+type TypeMetadata = {
+  ignoredFields?: Set<string>,
+  fieldMap?: { [string]: ValueDescriptor },
+  typeName?: string,
+  dirty?: boolean, // tracks structural changes only
+}
+
+type Count = Number
```
Comment (muescha) [inline packages/gatsby/src/schema/infer/inference-metadata.js]: is this here a typo with uppercase?

```suggestion
type Count = number
```

Comment (vladar) [reply packages/gatsby/src/schema/infer/inference-metadata.js]: Yep, thanks for noticing!

File: packages/gatsby/src/utils/gatsby-webpack-eslint-config-reload.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  constructor(options) {
+    this.plugin = { name: `GatsbyWebpackEslintConfigReload` }
+    this.options = options || {}
+  }
+
+  findEslintOptions = compiler =>
+    compiler.options.module.rules.find(isEslintRule).use[0].options
+
+  apply(compiler) {
+    compiler.hooks.compile.tap(this.plugin, stats => {
```
Comment (wardpeet) [inline packages/gatsby/src/utils/gatsby-webpack-eslint-config-reload.js]: This feels like it should be the string and not the object but I might be wrong.
```suggestion
    compiler.hooks.compile.tap(this.plugin.name, stats => {
```

Comment (vladar) [reply packages/gatsby/src/utils/gatsby-webpack-eslint-config-reload.js]: You are right, thanks for noticing!

Comment (vladar) [reply packages/gatsby/src/utils/gatsby-webpack-eslint-config-reload.js]: I used `GatsbyWebpackStatsExtractor` as a prototype for this plugin and it passes an object. But webpack docs say it should be a string. I'll change both plugins then.

File: packages/gatsby/src/utils/webpack-utils.js [hunk 6]
```diff
@@ -623,6 +624,9 @@ module.exports = async ({
 
   plugins.extractStats = options => new GatsbyWebpackStatsExtractor(options)
 
+  plugins.reloadESLintConfig = options =>
```
[the remark this answers is not in the document]
Comment (vladar) [reply packages/gatsby/src/utils/webpack-utils.js]: Renamed, thanks!

File: packages/gatsby/src/utils/webpack-utils.js [hunk 7]
```diff
@@ -8,6 +8,7 @@ const OptimizeCssAssetsPlugin = require(`optimize-css-assets-webpack-plugin`)
 const isWsl = require(`is-wsl`)
 
 const GatsbyWebpackStatsExtractor = require(`./gatsby-webpack-stats-extractor`)
+const GatsbyWebpackEslintConfigReload = require(`./gatsby-webpack-eslint-config-reload`)
```
[the remark this answers is not in the document]
Comment (vladar) [reply packages/gatsby/src/utils/webpack-utils.js]: Done

**Comment by vladar** [comment]
@wardpeet I tried doing webpack invalidation but for some reason, eslint-loader is not running after it. For now, we re-run queries here:

https://github.com/gatsbyjs/gatsby/blob/55ca80e317ecbca2f566072c929d5afe9a10b897/packages/gatsby/src/bootstrap/schema-hot-reloader.js#L29-L30

which will fail with error anyway. I will add invalidation to possible follow-ups and will probably need your help to debug why it is not working as expected.

**Comment by wardpeet** [comment]
Sweet! Sounds great, I'm not 100% sure my pseudo-code was the right call. Happy to debug later :)

**Comment by AileenCGN** [comment]
Hey @vladar 👋🏼
I have a question re this task:

>  Test with multiple medium to big size projects

Has this been done actually? I'm asking because the jump to Gatsby@2.18.0 significantly slowed down the development process for us:

With Gastby@2.18.0 after changing some markup in a page, `createPages` takes almost **30s**:
```bash
success Building development bundle - 22.864s
info added file at /Users/aileen/code/G3/src/pages/members.js
success write out requires - 0.007s
success createPages - 28.390s
success run queries - 0.104s - 0/1 9.58/s
success extract queries from components - 2.147s
success write out requires - 0.007s
success Re-building development bundle - 3.067s
success run queries - 4.639s - 385/385 82.99/s
```

With Gatsby@2.17.17 after changing some markup in a page, `createPages` takes **4s**:
```bash
success Building development bundle - 21.763s
info added file at /Users/aileen/code/G3/src/pages/members.js
success write out requires - 0.006s
success createPages - 3.962s
success run queries - 0.105s - 0/1 9.54/s
success extract queries from components - 2.090s
success write out requires - 0.004s
success Re-building development bundle - 3.149s
success run queries - 4.879s - 385/385 78.90/s
success run queries - 0.078s - 10/10 127.74/s
```

Is this known? Do you want me to open an issue for that? For now, we'll have to stick using the pre 2.18 version, as developing is kinda unbearable otherwise.

**Comment by vladar** [comment]
There are two known performance regressions. One in 2.18.0 after schema rebuilding and another one in 2.18.1 after https://github.com/gatsbyjs/gatsby/pull/17681

I guess we need to figure out which one affects you the most. Could you maybe try explicit version 2.18.0 and 2.18.1 and compare results?

**Comment by AileenCGN** [comment]
Oh!! Good call! Sorry, for not being accurate enough 😬

Seems like the longer rebuilding time is actually caused by 2.18.1 👇🏼

With Gatsby@2.18.0 (~3.2s):
```bash
info added file at /Users/aileen/code/G3/src/pages/members.js
success write out requires - 0.003s
success createPages - 3.158s
success run queries - 0.103s - 1/1 9.69/s
success extract queries from components - 1.753s
success write out requires - 0.006s
success Re-building development bundle - 2.537s
success run queries - 4.605s - 385/385 83.60/s
success run queries - 0.060s - 10/10 165.95/s
```

With Gatsby@2.18.1 (~21.8s):
```bash
info added file at /Users/aileen/code/G3/src/pages/members.js
success extract queries from components - 0.201s
success write out requires - 0.006s
success createPages - 21.790s
success run queries - 0.118s - 1/1 8.46/s
success write out requires - 0.004s
success Re-building development bundle - 22.950s
success run queries - 4.647s - 385/385 82.84/s
```

**Comment by vladar** [comment]
@AileenCGN Potential fix for this regression is published in `gatsby 2.18.4` (see PR #19774). Could you try again and maybe post here if it improves things for you?

**Comment by AileenCGN** [comment]
@vladar Awesome!! It's fixed now 🎉 Back to old rebuild times now 🤗

**Comment by tu4mo** [comment]
Hey @vladar, this seems to have somehow broken gatsby-transform-react-docgen, see #20043.

**Pull request merged**